In [3]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [15]:
pd.set_option('display.max_columns', None)

In [23]:
file_path = "../data/homicidios.xlsx"
df = pd.read_excel(file_path, sheet_name="HECHOS")


Realizaremos el proceso de limpieza teniendo en cuenta las situaciones más comunes:

1. Datos faltantes en algunas celdas
2. Columnas irrelevantes (que no responden al problema que queremos resolver)
3. Registros (filas) repetidos
4. Valores extremos (*outliers*) en el caso de las variables numéricas.

Al final de este proceso de limpieza deberíamos tener un set de datos **íntegro**, listo para la fase de Análisis Exploratorio.

In [4]:
print(df.shape)
df.head()

(696, 21)


ID  N_VICTIMAS      FECHA  AAAA  MM  DD      HORA HH  \
0  2016-0001           1 2016-01-01  2016   1   1  04:00:00  4   
1  2016-0002           1 2016-01-02  2016   1   2  01:15:00  1   
2  2016-0003           1 2016-01-03  2016   1   3  07:00:00  7   
3  2016-0004           1 2016-01-10  2016   1  10  00:00:00  0   
4  2016-0005           1 2016-01-21  2016   1  21  05:20:00  5   

                             LUGAR_DEL_HECHO TIPO_DE_CALLE  ...  Altura  \
0  AV PIEDRA BUENA Y AV FERNANDEZ DE LA CRUZ       AVENIDA  ...     NaN   
1          AV GRAL PAZ  Y AV DE LOS CORRALES      GRAL PAZ  ...     NaN   
2                         AV ENTRE RIOS 2034       AVENIDA  ...  2034.0   
3      AV LARRAZABAL Y GRAL VILLEGAS CONRADO       AVENIDA  ...     NaN   
4  AV SAN JUAN Y PRESIDENTE LUIS SAENZ PEÑA        AVENIDA  ...     NaN   

                                 Cruce  \
0  FERNANDEZ DE LA CRUZ, F., GRAL. AV.   
1                  DE LOS CORRALES AV.   
2                                  NaN   
3             VILLEGAS, CONRADO, GRAL.   
4              SAENZ PE?A, LUIS, PRES.   

                               Dirección Normalizada COMUNA  \
0  PIEDRA BUENA AV. y FERNANDEZ DE LA CRUZ, F., G...      8   
1               PAZ, GRAL. AV. y DE LOS CORRALES AV.      9   
2                                ENTRE RIOS AV. 2034      1   
3          LARRAZABAL AV. y VILLEGAS, CONRADO, GRAL.      8   
4            SAN JUAN AV. y SAENZ PEÑA, LUIS, PRES.      1   

                                 XY (CABA)         pos x         pos y  \
0    Point (98896.78238426 93532.43437792)  -58.47533969  -34.68757022   
1    Point (95832.05571093 95505.41641999)  -58.50877521  -34.66977709   
2   Point (106684.29090040 99706.57687843)  -58.39040293  -34.63189362   
3    Point (99840.65224780 94269.16534422)  -58.46503904  -34.68092974   
4  Point (106980.32827929 100752.16915795)  -58.38718297  -34.62246630   

    PARTICIPANTES VICTIMA    ACUSADO  
0       MOTO-AUTO    MOTO       AUTO  
1  AUTO-PASAJEROS    AUTO  PASAJEROS  
2       MOTO-AUTO    MOTO       AUTO  
3         MOTO-SD    MOTO         SD  
4  MOTO-PASAJEROS    MOTO  PASAJEROS  

[5 rows x 21 columns]

In [19]:
duplicados_totales = df.duplicated().sum()
duplicados_totales

0

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 696 entries, 0 to 695
Data columns (total 21 columns):
 #   Column                 Non-Null Count  Dtype         
---  ------                 --------------  -----         
 0   ID                     696 non-null    object        
 1   N_VICTIMAS             696 non-null    int64         
 2   FECHA                  696 non-null    datetime64[ns]
 3   AAAA                   696 non-null    int64         
 4   MM                     696 non-null    int64         
 5   DD                     696 non-null    int64         
 6   HORA                   696 non-null    object        
 7   HH                     696 non-null    object        
 8   LUGAR_DEL_HECHO        696 non-null    object        
 9   TIPO_DE_CALLE          696 non-null    object        
 10  Calle                  695 non-null    object        
 11  Altura                 129 non-null    float64       
 12  Cruce                  525 non-null    object        
 13  Direc

Los datos no están completos, pues no todas las columnas tienen la misma cantidad de registros.

El número total de registros debería ser 696. Sin embargo las columnas "Altura", "Cruce", "Dirección Normalizada" y "Calle" tienen datos faltantes.

Por ser un set de datos reducido, optaremos por imputar las filas correspondientes.

Revisaremos en primer lugar los registros faltantes en "Calle" y "Dirección Normalizada" para ver que ocurre en esos casos, considerando que es una cantidad pequeña.

In [12]:
# Filtrar registros donde la columna "Calle" es nula
calle_nula = df[df['Calle'].isnull()]

# Filtrar registros donde la columna "Direccion normalizada" es nula
direccion_nula = df[df['Dirección Normalizada'].isnull()]

In [16]:
calle_nula

ID  N_VICTIMAS      FECHA  AAAA  MM  DD      HORA  HH  \
119  2016-0151           1 2016-11-18  2016  11  18  20:35:00  20   

    LUGAR_DEL_HECHO TIPO_DE_CALLE Calle  Altura Cruce Dirección Normalizada  \
119              SD         CALLE   NaN     NaN   NaN                   NaN   

     COMUNA    XY (CABA) pos x pos y PARTICIPANTES VICTIMA ACUSADO  
119       0  Point (. .)     .     .     PEATON-SD  PEATON      SD

In [17]:
direccion_nula

ID  N_VICTIMAS      FECHA  AAAA  MM  DD      HORA  HH  \
38   2016-0052           1 2016-04-20  2016   4  20  20:00:00  20   
106  2016-0136           1 2016-10-25  2016  10  25  00:00:00   0   
119  2016-0151           1 2016-11-18  2016  11  18  20:35:00  20   
180  2017-0050           2 2017-04-28  2017   4  28  11:08:08  11   
181  2017-0051           1 2017-05-01  2017   5   1  03:47:47   3   
313  2018-0039           1 2018-04-21  2018   4  21  22:15:00  22   
546  2020-0026           1 2020-05-17  2020   5  17  06:40:00   6   
621  2021-0023           1 2021-03-01  2021   3   1  09:20:00   9   

                             LUGAR_DEL_HECHO TIPO_DE_CALLE  \
38                AUTOPISTA LUGONES PK 10000     AUTOPISTA   
106         AU BUENOS AIRES - LA PLATA KM. 4     AUTOPISTA   
119                                       SD         CALLE   
180  AU PERITO MORENO Y RAMAL ENLACE AU1/AU6     AUTOPISTA   
181                       AU DELLEPIANE 2400     AUTOPISTA   
313                 AUTOPISTA LUGONES KM 4.7     AUTOPISTA   
546             LUGONES, LEOPOLDO AV. KM 6,1     AUTOPISTA   
621         AU BUENOS AIRES LA PLATA  KM 4,5     AUTOPISTA   

                                    Calle  Altura Cruce Dirección Normalizada  \
38                  LUGONES, LEOPOLDO AV.     NaN   NaN                   NaN   
106     AUTOPISTA BUENOS AIRES - LA PLATA     NaN   NaN                   NaN   
119                                   NaN     NaN   NaN                   NaN   
180               AUTOPISTA PERITO MORENO     NaN   NaN                   NaN   
181  AUTOPISTA DELLEPIANE LUIS TTE. GRAL.     NaN   NaN                   NaN   
313                 LUGONES, LEOPOLDO AV.     NaN   NaN                   NaN   
546                 LUGONES, LEOPOLDO AV.     NaN   NaN                   NaN   
621     AUTOPISTA BUENOS AIRES - LA PLATA     NaN   NaN                   NaN   

     COMUNA    XY (CABA) pos x pos y     PARTICIPANTES VICTIMA      ACUSADO  
38       13  Point (. .)     .     .           MOTO-SD    MOTO           SD  
106       4  Point (. .)     .     .       MOTO-CARGAS    MOTO       CARGAS  
119       0  Point (. .)     .     .         PEATON-SD  PEATON           SD  
180       9  Point (. .)     .     .       MOTO-CARGAS    MOTO       CARGAS  
181       7  Point (. .)     .     .         AUTO-AUTO    AUTO         AUTO  
313      14  Point (. .)     .     .       PEATON-AUTO  PEATON         AUTO  
546      14  Point (. .)     .     .  MOTO-OBJETO FIJO    MOTO  OBJETO FIJO  
621       4  Point (. .)     .     .       MOTO-CARGAS    MOTO       CARGAS

In [33]:
from fuzzywuzzy import fuzz

# Suponiendo que tu DataFrame se llama 'df'

def preprocesar_texto(texto):
    # Convertir a minúsculas y reemplazar "AV" por "AVENIDA"
    texto = str(texto).lower().replace("av", "avenida")
    # Otros pasos de preprocesamiento según sea necesario
    # ...

    return texto


# Función para calcular la similitud de Levenshtein con preprocesamiento
def calcular_similitud(row):
    lugar_hecho = preprocesar_texto(row['LUGAR_DEL_HECHO'])
    calle = preprocesar_texto(row['Calle'])
    cruce = preprocesar_texto(row['Cruce'])
    altura = preprocesar_texto(str(row['Altura']))

    # Calcular similitud para cada combinación
    similitud_lugar_calle = fuzz.ratio(lugar_hecho, calle)
    similitud_lugar_cruce = fuzz.ratio(lugar_hecho, cruce)
    similitud_lugar_altura = fuzz.ratio(lugar_hecho, altura)
    similitud_lugar_completo = fuzz.ratio(lugar_hecho, f"{calle} {cruce} {altura}")

    # Seleccionar el máximo
    max_similitud = max(similitud_lugar_calle, similitud_lugar_cruce, similitud_lugar_altura, similitud_lugar_completo)

    return max_similitud

# Aplicar la función a cada fila del DataFrame
df['similitud'] = df.apply(calcular_similitud, axis=1)

In [31]:
df

ID  N_VICTIMAS      FECHA  AAAA  MM  DD      HORA  HH  \
0    2016-0001           1 2016-01-01  2016   1   1  04:00:00   4   
1    2016-0002           1 2016-01-02  2016   1   2  01:15:00   1   
2    2016-0003           1 2016-01-03  2016   1   3  07:00:00   7   
3    2016-0004           1 2016-01-10  2016   1  10  00:00:00   0   
4    2016-0005           1 2016-01-21  2016   1  21  05:20:00   5   
..         ...         ...        ...   ...  ..  ..       ...  ..   
691  2021-0093           1 2021-12-13  2021  12  13  17:10:00  17   
692  2021-0094           1 2021-12-20  2021  12  20  01:10:00   1   
693  2021-0095           1 2021-12-30  2021  12  30  00:43:00   0   
694  2021-0096           1 2021-12-15  2021  12  15  10:30:00  10   
695  2021-0097           1 2021-11-18  2021  11  18  06:10:00   6   

                               LUGAR_DEL_HECHO TIPO_DE_CALLE  \
0    AV PIEDRA BUENA Y AV FERNANDEZ DE LA CRUZ       AVENIDA   
1            AV GRAL PAZ  Y AV DE LOS CORRALES      GRAL PAZ   
2                           AV ENTRE RIOS 2034       AVENIDA   
3        AV LARRAZABAL Y GRAL VILLEGAS CONRADO       AVENIDA   
4    AV SAN JUAN Y PRESIDENTE LUIS SAENZ PEÑA        AVENIDA   
..                                         ...           ...   
691                          AV. RIESTRA Y MOM       AVENIDA   
692                    AU DELLEPIANE Y LACARRA     AUTOPISTA   
693                        AV. GAONA Y TERRADA       AVENIDA   
694                         AV. EVA PERON 4071       AVENIDA   
695                    PADRE CARLOS MUJICA 709         CALLE   

                            Calle  Altura  \
0                PIEDRA BUENA AV.     NaN   
1                  PAZ, GRAL. AV.     NaN   
2                  ENTRE RIOS AV.  2034.0   
3                  LARRAZABAL AV.     NaN   
4                    SAN JUAN AV.     NaN   
..                            ...     ...   
691                   RIESTRA AV.     NaN   
692  DELLEPIANE, LUIS, TTE. GRAL.     NaN   
693                     GAONA AV.     NaN   
694                PERON, EVA AV.  4071.0   
695           PADRE CARLOS MUJICA   709.0   

                                   Cruce  \
0    FERNANDEZ DE LA CRUZ, F., GRAL. AV.   
1                    DE LOS CORRALES AV.   
2                                    NaN   
3               VILLEGAS, CONRADO, GRAL.   
4                SAENZ PE?A, LUIS, PRES.   
..                                   ...   
691                                  MOM   
692                          LACARRA AV.   
693                              TERRADA   
694                                  NaN   
695                                  NaN   

                                 Dirección Normalizada  COMUNA  \
0    PIEDRA BUENA AV. y FERNANDEZ DE LA CRUZ, F., G...       8   
1                 PAZ, GRAL. AV. y DE LOS CORRALES AV.       9   
2                                  ENTRE RIOS AV. 2034       1   
3            LARRAZABAL AV. y VILLEGAS, CONRADO, GRAL.       8   
4              SAN JUAN AV. y SAENZ PEÑA, LUIS, PRES.       1   
..                                                 ...     ...   
691                                  RIESTRA AV. y MOM       7   
692         DELLEPIANE, LUIS, TTE. GRAL. y LACARRA AV.       9   
693                                GAONA AV. y TERRADA      11   
694                                PERON, EVA AV. 4071       9   
695                            PADRE CARLOS MUGICA 709       1   

                                   XY (CABA)         pos x         pos y  \
0      Point (98896.78238426 93532.43437792)  -58.47533969  -34.68757022   
1      Point (95832.05571093 95505.41641999)  -58.50877521  -34.66977709   
2     Point (106684.29090040 99706.57687843)  -58.39040293  -34.63189362   
3      Point (99840.65224780 94269.16534422)  -58.46503904  -34.68092974   
4    Point (106980.32827929 100752.16915795)  -58.38718297  -34.62246630   
..                                       ...           ...           ...   
691   Point (102728